### ISSUE TITLE
Autograd retain_graph=True bugs

### ISSUE THREAD

**Issue by leezu** [apache/incubator-mxnet#7268]
Consider the following example

```
import mxnet as mx
from mxnet import autograd
from mxnet import gluon

encoder = gluon.rnn.LSTM(hidden_size=300, num_layers=1)
encoder.collect_params().initialize(mx.init.Xavier(), ctx=mx.cpu())
decoder = gluon.rnn.LSTM(hidden_size=300, num_layers=1)
decoder.collect_params().initialize(mx.init.Xavier(), ctx=mx.cpu())

encoder_begin_state = encoder.begin_state(
    func=mx.nd.zeros, batch_size=8, ctx=mx.cpu())

expected_label = mx.nd.ones((8, 8))

loss = gluon.loss.SoftmaxCrossEntropyLoss()
# Encoder
with autograd.record():
    output, hidden = encoder(mx.nd.ones((8, 8, 300)), encoder_begin_state)

    # for i in hidden:
    #     i.attach_grad()
    hidden_detached = [i.detach() for i in hidden]
    for i in hidden_detached:
        i.attach_grad()

    prediction, _ = decoder(mx.nd.ones((8, 8, 300)), hidden_detached)
    l = loss(prediction, expected_label)

l.backward(retain_graph=True)
hidden[0].backward(hidden_detached[0].grad, retain_graph=True)
hidden[1].backward(hidden_detached[1].grad, retain_graph=True)

# Collect gradients to force execution
params = encoder.collect_params()
params.update(decoder.collect_params())
print([mx.nd.mean(p._grad[mx.cpu()]) for p in params.values()])
```

It will fail with `corrupted double-linked list`. Uncommenting
```
    # for i in hidden:
    #     i.attach_grad()
```
will fix that problem, but will stop the gradient from flowing back to the encoder parameters (which arguably should be documented?).

Calling backward only on hidden[0] or hidden[1] will also work.

Or is this not supposed to work? Essentially I want to decompose the autograd graph into two parts, similar to having two modules and using the input gradients of the second for the backward pass of the first.

@piiswrong

**Comment by piiswrong** [comment]
Your code works for me. Although the right way to do it is either
```
with autograd.record():
    output, hidden = encoder(mx.nd.ones((8, 8, 300)), encoder_begin_state)

    hidden_detached = [i for i in hidden]

    prediction, _ = decoder(mx.nd.ones((8, 8, 300)), hidden_detached)
    l = loss(prediction, expected_label)

l.backward()
```

or

```
with autograd.record():
    output, hidden = encoder(mx.nd.ones((8, 8, 300)), encoder_begin_state)

    hidden_detached = [i.detach() for i in hidden]
    for i in hidden_detached:
        i.attach_grad()

    prediction, _ = decoder(mx.nd.ones((8, 8, 300)), hidden_detached)
    l = loss(prediction, expected_label)

l.backward()
autograd.backward(hidden, [hidden_detached[0].grad, hidden_detached[1].grad])
```

**Comment by leezu** [comment]
Thanks. Using `autograd.backward` works for me. Your first proposed solution doesn't allow for making manual changes to the grad arrays before back-propagating.

What is your environment to run the sample code I provided? For me it reliably crashes mxnet 
```
*** Error in `python2': corrupted double-linked list: 0x000000000441a930 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x7bc02)[0x7fba1ef83c02]
/lib64/libc.so.6(+0x7c3e6)[0x7fba1ef843e6]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN5mxnet4exec16FComputeExecutorD1Ev+0xf9)[0x7fb996d3e169]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE10_M_releaseEv+0x132)[0x7fb995c6db42]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(+0x17e1bf2)[0x7fb996d16bf2]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(+0x1765959)[0x7fb996c9a959]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNSt17_Function_handlerIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEZNS0_6Engine8PushSyncESt8functionIFvS1_EENS0_7C
ontextERKSt6vectorIPNS2_3VarESaISC_EESG_NS0_10FnPropertyEiPKcEUlS1_S3_E_E9_M_invokeERKSt9_Any_dataS1_S3_+0x23)[0x7fb99687a173]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN5mxnet6engine14ThreadedEngine15ExecuteOprBlockENS_10RunContextEPNS0_8OprBlockE+0x97)[0x7fb996ca7357]
/home/data/lelausen/software/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNSt17_Function_handlerIFvvEZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS2_8OprBlockEbENKUlvE_clEvEUlvE_E9_M_in
vokeERKSt9_Any_data+0x42)[0x7fb996ca9e92]
/usr/local/lib/gcc4/lib64/libstdc++.so.6(+0xb3150)[0x7fb9fe61c150]
/lib64/libpthread.so.0(+0x7dc5)[0x7fba21d8adc5]
/lib64/libc.so.6(clone+0x6d)[0x7fba1efff76d]
```

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
